pkgs/build-support/dart/build-dart-application/default.nix

```diff
@@ -121,6 +121,30 @@ let
     extraSetupCommands = extraPackageConfigSetup;
   };
 
+  packageGraph = pub2nix.generatePackageGraph {
+    name =
+      if (args.pname != null && args.version != null) then "${args.pname}-${args.version}" else null;
+
+    dependencies = builtins.concatLists (builtins.attrValues pubspecLockData.dependencies);
+
+    inherit (pubspecLockData) dependencySources;
+
+    pubspecFile = lib.concatStringsSep "/" (
+      lib.filter (s: s != "") [
+        src
+        (lib.removePrefix "/" (
+          if lib.hasPrefix src.name sourceRoot then
+            lib.removePrefix src.name sourceRoot
+          else if lib.hasPrefix "source" sourceRoot then
+            lib.removePrefix "source" sourceRoot
+          else
+            sourceRoot
+        ))
+        "pubspec.yaml"
+      ]
+    );
```

Comment:
This is a little awkward. It also may be inaccurate if the `sourceRoot` _environment variable_ is modified at build time before the `unpackPhase`, though admittedly I haven't ever seen that happen.

It may be more appropriate to move the Python script invocation into a setup hook rather than a dedicated derivation. That way:

- The script can find `pubspec.yaml` in the current directory, without worrying about `sourceRoot`
- Any patches to `pubspec.yaml` by the `patchPhase` or custom scripts don't need to be duplicated (which is currently missing from this implementation too)